db/compaction_job.cc

```diff
@@ -747,6 +747,14 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
       shutting_down_));
   auto c_iter = sub_compact->c_iter.get();
   c_iter->SeekToFirst();
+  if (c_iter->Valid()) {
+    // ShouldStopBefore() maintains state based on keys processed so far. The
+    // compaction loop always calls it on the "next" key, thus won't tell it the
+    // first key. So we do that here.
+    bool should_stop = sub_compact->ShouldStopBefore(
```

Comment:
I didn't find the safety issue, but it's unnecessary to call it in the case you described, so I'll update the conditional.